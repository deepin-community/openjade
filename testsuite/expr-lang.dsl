<!doctype style-sheet PUBLIC "-//James Clark//DTD DSSSL Style Sheet//EN">
<!-- Copyright (C) 1999 Avi Kivity -->
                 
<![cdata[                
                 
(root
    (test-expression-language)
)

(define (test-expression-language)
    (sosofo-append
        (make-test "Character Primitives" test-char)
	(make-test "Character Properties" test-char-properties)
        (make-test "String Primitives" test-string)
	(make-test "Standard examples" test-examples)
    )
)

(define (test-char)
    (sosofo-append
        (testcase 'char=?      #\a #\b)
        (testcase 'char=?      #\a #\a)
        (testcase 'char=?      #\a #\A)
        (testcase 'char<?      #\a #\b)
        (testcase 'char<?      #\a #\a)
        (testcase 'char<?      #\a #\A)
        (testcase 'char<=?     #\a #\b)
        (testcase 'char<=?     #\a #\a)
        (testcase 'char<=?     #\a #\A)
        (testcase 'char>?      #\a #\b)
        (testcase 'char>?      #\a #\a)
        (testcase 'char>?      #\a #\A)
        (testcase 'char>=?     #\a #\b)
        (testcase 'char>=?     #\a #\a)
        (testcase 'char>=?     #\a #\A)
        (testcase 'char-ci=?   #\a #\b)
        (testcase 'char-ci=?   #\a #\a)
        (testcase 'char-ci=?   #\a #\A)
        (testcase 'char-ci<?   #\a #\b)
        (testcase 'char-ci<?   #\a #\a)
        (testcase 'char-ci<?   #\a #\A)
        (testcase 'char-ci<=?  #\a #\b)
        (testcase 'char-ci<=?  #\a #\a)
        (testcase 'char-ci<=?  #\a #\A)
        (testcase 'char-ci>?   #\a #\b)
        (testcase 'char-ci>?   #\a #\a)
        (testcase 'char-ci>?   #\a #\A)
        (testcase 'char-ci>=?  #\a #\b)
        (testcase 'char-ci>=?  #\a #\a)
        (testcase 'char-ci>=?  #\a #\A)
    )
)

(add-char-properties funny: #\a #\a)
(add-char-properties funny: (char-property 'funny #\a) #\b #\c)
(add-char-properties funny: (char-property 'funny #\z) #\d)
(declare-char-property funny 'boring)
(define (test-char-properties)
    (sosofo-append
        (testcase 'char-property 'space? #\space)
        (testcase 'char-property 'space? #\space 'foo)
	(testcase 'char-property 'space? #\exclamation-mark)
	(testcase 'char-property 'space? #\exclamation-mark 'foo)
        (testcase 'char-property 'break-before-priority #\space)
	(testcase 'char-property 'break-after-priority #\space)
	(testcase 'char-property 'break-before-priority #\a)
	(testcase 'char-property 'break-after-priority #\a)
	(testcase 'char-property 'funny #\a)
	(testcase 'char-property 'funny #\b)
	(testcase 'char-property 'funny #\c)
	(testcase 'char-property 'funny #\d)
	(testcase 'char-property 'funny #\e)
	(testcase 'char-property 'funny #\z)
    ))

(define (test-string)
    (empty-sosofo)
)

(define (test-examples)
    (sosofo-append
        (testcase 'not  #t)
        (testcase 'not  3)
        (testcase 'not  (list 3))
        (testcase 'not  #f)
        (testcase 'not  '())
        (testcase 'not  (list))
        (testcase 'not  'nil)
        (testcase 'boolean?  #f)
        (testcase 'boolean?  0)
        (testcase 'boolean?  '())
        (testcase 'pair?  '(a . b))
        (testcase 'pair?  '(a b c))
        (testcase 'pair?  '())
        (testcase 'cons  'a '())
        (testcase 'cons  '(a) '(b c d))
        (testcase 'cons  "a" '(b c))
        (testcase 'cons  'a 3)
        (testcase 'cons  '(a b) 'c)
	(testcase 'car  '(a b c))
	(testcase 'car  '((a) b c d))
	(testcase 'car  '(1 . 2))
	;(testcase 'car  '()) ;==> error
        (testcase 'cdr	'((a) b c d))
	(testcase 'cdr 	'(1 . 2))
	;(testcase 'cdr  '()) ;==> error
	(testcase 'list?  '(a b c))
	(testcase 'list?  '())
	(testcase 'list?  '(a . b))
	(testcase 'list  'a (+ 3 4) 'c)
	(testcase 'list)
	(testcase 'length  '(a b c))
	(testcase 'length  '(a (b) (c d e)))
	(testcase 'length  '())
	(testcase 'append  '(x) '(y))
	(testcase 'append  '(a) '(b c d))
	(testcase 'append  '(a (b)) '((c))) 	
	(testcase 'append  '(a b) '(c . d)) 	
	(testcase 'append  '() 'a) 	
	(testcase 'reverse  '(a b c))
	(testcase 'reverse  '(a (b c) d (e (f))))
	(testcase 'list-ref  '(a b c d) 2)
	(testcase 'list-ref  '(a b c d) (inexact->exact (round 1.8)))
	(testcase 'member  'a '(a b c))
	(testcase 'member  'b '(a b c))
	(testcase 'member  'a '(b c d))
	(testcase 'assoc  'a '((a 1) (b 2) (c 3)))
	(testcase 'assoc  'b '((a 1) (b 2) (c 3)))
	(testcase 'assoc  'd '((a 1) (b 2) (c 3)))
	(testcase 'symbol?  'foo)
	(testcase 'symbol?  (car '(a b)))
	(testcase 'symbol?  "bar")
	(testcase 'symbol?  'nil)
	(testcase 'symbol?  '())
	(testcase 'symbol?  #f)
	(testcase 'symbol->string  'flying-fish)
	(testcase 'symbol->string (string->symbol "Malvina"))
	(testcase 'equal?  'mISSISSIppi 'mississippi)
	(testcase 'equal?  'bitBlt (string->symbol "bitBlt"))
	(testcase 'equal?  'JollyWog 
		(string->symbol (symbol->string 'JollyWog)))
	(testcase 'string=? "K. Harper, M.D."
		(symbol->string (string->symbol "K. Harper, M.D.")))
	(testcase 'keyword->string  Argentina:)
	(testcase 'string->keyword  "foobar")
	(testcase 'real?  3)
	(testcase 'integer?  3.0)
	(testcase 'max  3 4)
	(testcase 'max  3.9 4)
	(testcase '+  3 4)
	(testcase '+  3)
	(testcase '+) 
	(testcase '*  4)
	(testcase '*)
	(testcase '-  3 4)
	(testcase '-  3 4 5)
	(testcase '-  3)
	(testcase '/  3 4 5)
	(testcase '/  3)
        (testcase 'abs  -7)
	(testcase 'modulo  13 4)
	(testcase 'remainder  13 4)
	(testcase 'modulo  -13 4)
	(testcase 'remainder  -13 4)
	(testcase 'modulo  13 -4)
	(testcase 'remainder  13 -4)
	(testcase 'modulo  -13 -4)
	(testcase 'remainder  -13 -4)
	(testcase 'remainder  -13 -4.0)
	(testcase 'floor  -4.3)
	(testcase 'ceiling  -4.3)
	(testcase 'truncate  -4.3)
	(testcase 'round  -4.3)
	(testcase 'floor  3.5)
	(testcase 'ceiling  3.5)
	(testcase 'truncate  3.5)
	(testcase 'round  3.5)
	(testcase 'round  7)
	(testcase 'string->number  "100")
	(testcase 'string->number  "100" 16)
	(testcase 'string->number  "1e2")
	(testcase 'procedure?  car)
	(testcase 'procedure?  'car)
	(testcase 'procedure?  (lambda (x) (* x x)))
	(testcase 'procedure?  '(lambda (x) (* x x)))
	(testcase 'apply  + (list 3 4))
	(testcase 'map  cadr '((a b) (d e) (g h)))
	(testcase 'map  (lambda (n) (expt n n)) '(1 2 3 4 5))
	(testcase 'map  + '(1 2 3) '(4 5 6))
    )
)

(define symbol-proc-map
    `(
        (char=? . ,char=?) (char<? . ,char<?) (char<=? . ,char<=?) 
        (char>? . ,char>?) (char>=? . ,char>=?) (char-ci=? . ,char-ci=?)
        (char-ci<? . ,char-ci<?) (char-ci<=? . ,char-ci<=?)
        (char-ci>? . ,char-ci>?) (char-ci>=? . ,char-ci>=?)
	(char-property . ,char-property)
        (not . ,not) (boolean? . ,boolean?) (pair? . ,pair?) 
        (cons . ,cons) (car . ,car) (cdr . ,cdr) (list? . ,list?)
	(list . ,list) (length . ,length) (append . ,append)
	(reverse . ,reverse) (list-ref . ,list-ref) (member . ,member)
	(assoc . ,assoc) (symbol? . ,symbol?) 
	(symbol->string . ,symbol->string) (equal? . ,equal?)
	(string=? . ,string=?) (keyword->string . ,keyword->string)
	(string->keyword . ,string->keyword) (real? . ,real?)
	(integer? . ,integer?) (max . ,max) (+ . ,+) (* . ,*) (- . ,-)
	(/ . ,/) (abs . ,abs) (modulo . ,modulo) (remainder . ,remainder)
	(floor . ,floor) (ceiling . ,ceiling) (truncate . ,truncate)
	(round . ,round) (string->number . ,string->number)
	(procedure? . ,procedure?) (apply . ,apply) (map . ,map)
    )     
)
     
(define (symbol->proc sym) (cdr (assoc sym symbol-proc-map)))     
                  
(define (testcase proc-sym #!rest args)
    (write-line
        (string-append
            "(" 
            (symbol->string proc-sym)
            (reduce 
                args
                (lambda (args arg) (string-append args " " (obj->string arg)))
                ""
            )
            ") --> "
            (obj->string (apply (symbol->proc proc-sym) args))
        )
    )
)       

(define (reduce list combine init)
  (let loop ((result init)
             (list list))
    (if (null? list)
        result
        (loop (combine result (car list))
              (cdr list)))))

(define (obj->string obj)
    (cond
        ((char? obj) (list->string (list #\# #\\ obj)))
        ((string? obj) (string-append "\"" obj "\""))
        ((boolean? obj) (if obj "#t" "#f"))
        ((symbol? obj) (symbol->string obj))
	((keyword? obj) (string-append (keyword->string obj) ":"))
        ((number? obj) (number->string obj)) 
	((procedure? obj) "procedure")
        ((list? obj) 
	    (if (null? obj)
		"()"
            	(apply string-append 
                    (append (list "(" (obj->string (car obj)))
                        (map (lambda (x) 
				(string-append " " (obj->string x))) (cdr obj))
                    '(")")))))
	((pair? obj) (string-append "(" (obj->string (car obj))
				" . " (obj->string (cdr obj)) ")"))
        (else "unimplemented")
    )
)           
                  
(define (make-test title proc)
    (sosofo-append
        (write-line title)
        (proc)
        (write-empty-line)
    )
)    

(define (write-empty-line) (write-line ""))

(define newline "\carriage-return;\line-feed;")

(define (write-line string)
    (make formatting-instruction
        data: (string-append string newline)
    )
)

(declare-flow-object-class formatting-instruction
  "UNREGISTERED::James Clark//Flow Object Class::formatting-instruction")

(declare-default-language (language "EN" "US"))

(define language
  (external-procedure "UNREGISTERED::OpenJade//Procedure::language"))

]]>
