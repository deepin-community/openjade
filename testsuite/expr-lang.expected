Character Primitives
(char=? #\a #\b) --> #f
(char=? #\a #\a) --> #t
(char=? #\a #\A) --> #f
(char<? #\a #\b) --> #t
(char<? #\a #\a) --> #f
(char<? #\a #\A) --> #f
(char<=? #\a #\b) --> #t
(char<=? #\a #\a) --> #t
(char<=? #\a #\A) --> #f
(char>? #\a #\b) --> #f
(char>? #\a #\a) --> #f
(char>? #\a #\A) --> #t
(char>=? #\a #\b) --> #f
(char>=? #\a #\a) --> #t
(char>=? #\a #\A) --> #t
(char-ci=? #\a #\b) --> #f
(char-ci=? #\a #\a) --> #t
(char-ci=? #\a #\A) --> #t
(char-ci<? #\a #\b) --> #t
(char-ci<? #\a #\a) --> #f
(char-ci<? #\a #\A) --> #f
(char-ci<=? #\a #\b) --> #t
(char-ci<=? #\a #\a) --> #t
(char-ci<=? #\a #\A) --> #t
(char-ci>? #\a #\b) --> #f
(char-ci>? #\a #\a) --> #f
(char-ci>? #\a #\A) --> #f
(char-ci>=? #\a #\b) --> #f
(char-ci>=? #\a #\a) --> #t
(char-ci>=? #\a #\A) --> #t

Character Properties
(char-property space? #\ ) --> #t
(char-property space? #\  foo) --> #t
(char-property space? #\!) --> #f
(char-property space? #\! foo) --> foo
(char-property break-before-priority #\ ) --> 2
(char-property break-after-priority #\ ) --> 3
(char-property break-before-priority #\a) --> 1
(char-property break-after-priority #\a) --> 1
(char-property funny #\a) --> #\a
(char-property funny #\b) --> #\a
(char-property funny #\c) --> #\a
(char-property funny #\d) --> boring
(char-property funny #\e) --> boring
(char-property funny #\z) --> boring

String Primitives

Standard examples
(not #t) --> #f
(not 3) --> #f
(not (3)) --> #f
(not #f) --> #t
(not ()) --> #f
(not ()) --> #f
(not nil) --> #f
(boolean? #f) --> #t
(boolean? 0) --> #f
(boolean? ()) --> #f
(pair? (a . b)) --> #t
(pair? (a b c)) --> #t
(pair? ()) --> #f
(cons a ()) --> (a)
(cons (a) (b c d)) --> ((a) b c d)
(cons "a" (b c)) --> ("a" b c)
(cons a 3) --> (a . 3)
(cons (a b) c) --> ((a b) . c)
(car (a b c)) --> a
(car ((a) b c d)) --> (a)
(car (1 . 2)) --> 1
(cdr ((a) b c d)) --> (b c d)
(cdr (1 . 2)) --> 2
(list? (a b c)) --> #t
(list? ()) --> #t
(list? (a . b)) --> #f
(list a 7 c) --> (a 7 c)
(list) --> ()
(length (a b c)) --> 3
(length (a (b) (c d e))) --> 3
(length ()) --> 0
(append (x) (y)) --> (x y)
(append (a) (b c d)) --> (a b c d)
(append (a (b)) ((c))) --> (a (b) (c))
(append (a b) (c . d)) --> (a . (b . (c . d)))
(append () a) --> a
(reverse (a b c)) --> (c b a)
(reverse (a (b c) d (e (f)))) --> ((e (f)) d (b c) a)
(list-ref (a b c d) 2) --> c
(list-ref (a b c d) 2) --> c
(member a (a b c)) --> (a b c)
(member b (a b c)) --> (b c)
(member a (b c d)) --> #f
(assoc a ((a 1) (b 2) (c 3))) --> (a 1)
(assoc b ((a 1) (b 2) (c 3))) --> (b 2)
(assoc d ((a 1) (b 2) (c 3))) --> #f
(symbol? foo) --> #t
(symbol? a) --> #t
(symbol? "bar") --> #f
(symbol? nil) --> #t
(symbol? ()) --> #f
(symbol? #f) --> #f
(symbol->string flying-fish) --> "flying-fish"
(symbol->string Malvina) --> "Malvina"
(equal? mISSISSIppi mississippi) --> #f
(equal? bitBlt bitBlt) --> #t
(equal? JollyWog JollyWog) --> #t
(string=? "K. Harper, M.D." "K. Harper, M.D.") --> #t
(keyword->string Argentina:) --> "Argentina"
(string->keyword "foobar") --> foobar:
(real? 3) --> #t
(integer? 3) --> #t
(max 3 4) --> 4
(max 3.9 4) --> 4
(+ 3 4) --> 7
(+ 3) --> 3
(+) --> 0
(* 4) --> 4
(*) --> 1
(- 3 4) --> -1
(- 3 4 5) --> -6
(- 3) --> -3
(/ 3 4 5) --> 0.15
(/ 3) --> 0.333333
(abs -7) --> 7
(modulo 13 4) --> 1
(remainder 13 4) --> 1
(modulo -13 4) --> 3
(remainder -13 4) --> -1
(modulo 13 -4) --> -3
(remainder 13 -4) --> 1
(modulo -13 -4) --> -1
(remainder -13 -4) --> -1
(remainder -13 -4) --> -1
(floor -4.3) --> -5
(ceiling -4.3) --> -4
(truncate -4.3) --> -4
(round -4.3) --> -4
(floor 3.5) --> 3
(ceiling 3.5) --> 4
(truncate 3.5) --> 3
(round 3.5) --> 4
(round 7) --> 7
(string->number "100") --> 100
(string->number "100" 16) --> 256
(string->number "1e2") --> 100
(procedure? procedure) --> #t
(procedure? car) --> #f
(procedure? procedure) --> #t
(procedure? (lambda (x) (* x x))) --> #f
(apply procedure (3 4)) --> 7
(map procedure ((a b) (d e) (g h))) --> (b e h)
(map procedure (1 2 3 4 5)) --> (1 4 27 256 3125)
(map procedure (1 2 3) (4 5 6)) --> (5 7 9)

